### ISSUE TITLE
[shapes] Draw circle

### ISSUE THREAD

**Issue by lofcz** [MonoGame-Extended/Monogame-Extended#574]
Noticed a bug with `DrawCircle`. In the image below is a circle of radius 32, drawn with red color, half opacity (alpha = .5f) and 32 sides. The issue should be clear to see - rather than drawing the circle with half opacity, some triangles from which the fan is made of are overlapping, resulting in the effect below. On the other hand, functions work fine with alpha = 1. Please note, that some other MonoGame.Extended shapes (ex. rectangle) work fine. I'd be interested in fixing this issue myself, if you approve this is unintended behavior. 

Edit: would be nice to have DrawTriangle, FillCircle & FillTriangle. 
![noname](https://user-images.githubusercontent.com/10260230/49024864-bcd2f680-f19a-11e8-9cd4-42dd366547ac.jpg)

**Comment by craftworkgames** [comment]
> if you approve this is unintended behavior.

Ah yes, please do. 

Oh btw, while we are on the topic I previously got permission to borrow the [DebugView.cs](https://github.com/VelcroPhysics/VelcroPhysics/blob/master/DebugViews/MonoGame/DebugView.cs#L567) code from the VelcroPhysics library (formerly known as Farseer Physics). 

I believe their shape algorithm's are better than ours and they use a `PrimitiveBatch` class that draws the shapes in a cleaner way (it doesn't rely on `SpriteBatch`).

**Comment by lofcz** [comment]
Actually found that the bug is present in all shapes. Reproduce:
Change background to black
Set alpha to 0.5f
Draw rectangle 
Notice how joint overlap

For myself I've solved this by drawing shapes with 3d textures vertices, using basic effect, provided with matrices (world, view, projection) to represent 3d world equal to camera 2d view matrix.

**Comment by craftworkgames** [comment]
> drawing shapes with 3d color vertices, using basic effect

Yeah, that's how Velcro's `PrimitiveBatch` does it I believe. I think that's the right way to go. The only downside I can think of is that you can't draw a shape in the middle of another `SpriteBatch` call without messing around with the depth. 

On the other hand, the `PrimitiveBatch` could also be modified to render sprites so maybe it's not a big deal either way.

**Comment by lofcz** [comment]
Update: Personally I found no way around this to preserve one batch call for everything. I was really scared about performance drops in case of multiple batches but it turned up only about 5-7% for me. 
![noname](https://user-images.githubusercontent.com/10260230/49158720-19f9b400-f323-11e8-9e4c-87e76932e26d.png)

**Comment by lofcz** [comment]
**Update** - worked out entire shapes library, currently optimizing via object pooling. 
https://github.com/lofcz/SimplexRpgEngine/blob/master/SimplexCore/SgmlDraw.cs

**Comment by lithiumtoast** [comment]
It seems the solution is to upgrade the code to support drawing shapes using a custom batcher.

**Comment by AristurtleDev** [comment]
While looking into this issue, it seems there were two bugs that were causing the original problem:  Fixing the first one revealed the second one.

1. `DrawSolidCircle`, `DrawSolidPolygon`, and `DrawSolidEllipse` in `PrimitiveDrawing` are silently multiplying the caller's color by `0.5f` internally. This means passing something like `new Color(255, 0, 0, 128)` didn't give 50% alpha, it gave 25% and the caller had no way to get the color they actually wanted
2. `PrimitiveBatch.Begin()` never set a `BlendState` so the GPU was using whatever state was left over from the previous draw call.  This seemed to often be `BlendState.Opaque` after `GraphicsDevice.Clear()` which silently discards alpha entirely

Concerning the broader conversation that happened here, MonoGame.Extended has two distinct paths for drawing shapes:

- `ShapeExtensions`: SpriteBatch based and outline only.  These are fine for debug and prototype work, but has known issues with join-overlap artifacts at corners when using semi-transparent color. Additionally solid filled shapes with correct alpha are fundamentally impossible this way (SpriteBatch can draw a dis as non-overlapping triangles in a single draw call)
- `PrimtiveDrawing` and `PrimitiveBatch`: GPU triangle lists using `BasicEffect`.  This is the corrct path for more production quality filled shapes with proper alpha blending.

A third option, `Batcher2D`.  This is something that @lithiumtoast originally began implementing but it was never completed, as acknowledged here https://github.com/MonoGame-Extended/Monogame-Extended/issues/530#issuecomment-407250057

> I'm not sure what the state is of Batcher2D right now. I originally wrote the code to mimic SpriteBatch but could also render shapes such as rectangles, circles, arcs, polygons, etc. SpriteBatch is very well optimized and recommended compared to Batcher2D

After reviewing it, it's defiitely not a finished feature, and also has no dependencies in the code base.  To fully implement this, it would need to have at minimum 1:1 feature parity with MonoGame's existing `SpriteBatch` in order to be promoted as a replacement that can also draw shapes.  Given that the currently implementation was never finished and the time investment this would take, i'm opting to instead mark that entire API point as obsolete and will be removing it in 6.0.0 release.

**Comment by lofcz** [comment]
Time flies, eh? Thanks for looking into this!

**Comment by AristurtleDev** [comment]
> Time flies, eh? Thanks for looking into this!

Only took what? 8 years 😅 
Trying to get through a lot of the older not-tiled ones haha

**Comment by vchelaru** [comment]
For future readers, I've been working with Apos.Shapes a lot lately and it's working well. This is the focus of the library and is well maintained so I recommend using that for anything beyond quick debugging:

https://github.com/Apostolique/Apos.Shapes
